lib/ansible/modules/windows/win_rds_cap.ps1

```diff
@@ -0,0 +1,323 @@
+$redirect_drives = Get-AnsibleParam -obj $params -name "redirect_drives" -type "bool"
+$redirect_printers = Get-AnsibleParam -obj $params -name "redirect_printers" -type "bool"
+$redirect_serial = Get-AnsibleParam -obj $params -name "redirect_serial" -type "bool"
+$redirect_pnp = Get-AnsibleParam -obj $params -name "redirect_pnp" -type "bool"
+
+
+function Get-CAP([string] $name) {
+    $cap_path = "RDS:\GatewayServer\CAP\$name"
+    $cap = @{
+        Name = $name
+    }
+
+    # Fetch CAP properties
+    Get-ChildItem -Path "$cap_path" | foreach { $cap.Add($_.Name,$_.CurrentValue) }
+    # Convert boolean values
+    $cap.Enabled = $cap.Status -eq 1
+    $cap.Remove("Status")
+    $cap.AllowOnlySDRTSServers = $cap.AllowOnlySDRTSServers -eq 1
+
+    # Convert multiple choices values
+    $cap.AuthMethod = $auth_methods_set[$cap.AuthMethod]
+    $cap.SessionTimeoutAction = $session_timeout_actions_set[$cap.SessionTimeoutAction]
+
+    # Fetch CAP device redirection settings
+    $cap.DeviceRedirection = @{}
+    Get-ChildItem -Path "$cap_path\DeviceRedirection" | foreach { $cap.DeviceRedirection.Add($_.Name, ($_.CurrentValue -eq 1)) }
+
+    # Fetch CAP user and computer groups
+    $cap.UserGroups = @(Get-ChildItem -Path "$cap_path\UserGroups" | Select -ExpandProperty Name)
+    $cap.ComputerGroups = @(Get-ChildItem -Path "$cap_path\ComputerGroups" | Select -ExpandProperty Name)
+
+    return $cap
+}
+
+function Set-CAPPropertyValue {
+    [CmdletBinding(SupportsShouldProcess=$true)]
+    param (
+        [Parameter(Mandatory=$true)]
+        [string] $name,
+        [Parameter(Mandatory=$true)]
+        [string] $property,
+        [Parameter(Mandatory=$true)]
+        $value,
+        [Parameter()]
+        $resultobj = @{}
+    )
+
+    $cap_path = "RDS:\GatewayServer\CAP\$name"
+
+    try {
+        Set-Item -Path "$cap_path\$property" -Value $value -ErrorAction Stop
+    } catch {
+        Fail-Json -obj $resultobj -message "Failed to set property $property of CAP ${name}: $($_.Exception.Message)"
+    }
+}
+
+$result = @{
+  changed = $false
+}
+$diff_text = $null
+
+# Validate CAP name
+if ($name -match "[*/\\;:?`"<>|\t]+") {
+    Fail-Json -obj $result -message "Invalid character in CAP name."
+}
+
+# Validate user groups
+if ($null -ne $user_groups) {
+    if ($user_groups.Count -lt 1) {
+        Fail-Json -obj $result -message "Parameter 'user_groups' cannot be an empty list."
+    }
+
+    $user_groups = $user_groups | foreach {
+        $group = $_
+        # Test that the group is resolvable on the local machine
+        $sid = Convert-ToSID -account_name $group
+        if (!$sid) {
+            Fail-Json -obj $result -message "$group is not a valid user group on the host machine or domain"
+        }
+
+        # Return the normalized group name in UPN format
+        $group_name = Convert-FromSID -sid $sid
+        ($group_name -split "\\")[1..0] -join "@"
```

Comment:
This doesn't seem right, e.g. `DOMAIN\User` would turn into `User@DOMAIN` when the domain part could really be `DOMAIN.ORG` or something like that. You might need cast the `$sid` string to a SecurityIdentifier object and find a better way to get the UPN from there. Why are you converting to the UPN in the first place?

If you are just using it to get a comparison string then you should be comparing the SIDs themselves.

Reply:
No it's not for comparison, the UPN form is required for the RDS PSProvider:
```powershell
# This doesn't seem to work if $group is not in UPN form
New-Item -Path "RDS:\GatewayServer\CAP\Test_CAP\UserGroups" -Name $group
```
And that gets on my nerve, because I know this is quick and dirty conversion but I didn't really find a better way to do it. [Through .NET, there is the DirectoryEntry object](https://stackoverflow.com/a/7524210) but it seems to rely on directory services, if I understand correctly.
If we don't find a better way, maybe we could either use WMI instead of PSProvider, or say that the UPN form is the one recommended for the parameter, and the others are supported for simple cases only...

Reply:
Let’s find the proper API to configure this and not really on editing the registry directly. If it requires DirectoryServices then that’s not the end of the world, it’s just an Add-Type cmdlet away.

Reply:
I havn't been able to find a simple and reliable way to get the group UPN from its SID, working for both local groups and domain groups, so I decide to use WMI methods instead.